terraform/context_apply_test.go

```diff
@@ -11122,3 +11122,63 @@ resource "aws_instance" "cbd" {
 		t.Fatal("aws_instance.foo should also be create_before_destroy")
 	}
 }
+
+func TestContext2Apply_moduleDependsOn(t *testing.T) {
+	m := testModule(t, "apply-module-depends-on")
+
+	p := testProvider("test")
+	p.ReadDataSourceResponse = providers.ReadDataSourceResponse{
+		State: cty.ObjectVal(map[string]cty.Value{
+			"id":  cty.StringVal("data"),
+			"foo": cty.NullVal(cty.String),
+		}),
+	}
+	p.DiffFn = testDiffFn
+
+	// each instance being applied should happen in sequential order
+	applied := int64(0)
+
+	p.ApplyResourceChangeFn = func(req providers.ApplyResourceChangeRequest) (resp providers.ApplyResourceChangeResponse) {
+		state := req.PlannedState.AsValueMap()
+		num, _ := state["num"].AsBigFloat().Float64()
+		ord := int64(num)
+		if !atomic.CompareAndSwapInt64(&applied, ord-1, ord) {
+			actual := atomic.LoadInt64(&applied)
+			resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("instance %d was applied after %d", ord, actual))
+		}
+
+		state["id"] = cty.StringVal(fmt.Sprintf("test_%d", ord))
+		resp.NewState = cty.ObjectVal(state)
+
+		return resp
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Config: m,
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+	})
+
+	_, diags := ctx.Plan()
+	if diags.HasErrors() {
+		t.Fatal(diags.ErrWithWarnings())
+	}
+
+	_, diags = ctx.Apply()
+	if diags.HasErrors() {
+		t.Fatal(diags.ErrWithWarnings())
+	}
+
+	// run the plan again to ensure that data sources are not going to be re-read
+	plan, diags := ctx.Plan()
```

Comment:
I might be missing something obvious, but I don't understand this comment. Why does re-planning after apply ensure that the data source isn't re-read?

Reply:
If the data source is read during plan (which is a new ability from #24904), it would show in the plan as a read operation. There should be nothing left to apply at this point, so anything in the plan should be NoOp.

Reply:
Ah, I see now, thanks!